src/ParallelAlgorithms/Amr/Criteria/TruncationError.hpp

```diff
@@ -0,0 +1,113 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <pup.h>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "Domain/Amr/Flag.hpp"
+#include "Domain/Tags.hpp"
+#include "NumericalAlgorithms/LinearOperators/PowerMonitors.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Options/String.hpp"
+#include "ParallelAlgorithms/Amr/Criteria/Criterion.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+template <size_t>
+class ElementId;
+/// \endcond
+
+namespace amr::Criteria {
+
+/*!
+ * \brief Refine the grid towards the target truncation error
+ *
+ * If any tensor component has a truncation error above the target value, the
+ * element will be p-refined. For details on how the truncation error is
+ * computed see `PowerMonitors::truncation_error`.
+ *
+ * \tparam Dim Spatial dimension of the grid
+ * \tparam TensorTags List of tags of the tensors to be monitored
+ */
+template <size_t Dim, typename TensorTags>
+class TruncationError : public Criterion {
+ public:
+  struct TargetTruncationError {
+    static std::string name() { return "Target"; }
+    using type = double;
+    static constexpr Options::String help = {
+        "The absolute target truncation error. If any tensor component "
+        "has a truncation error above this value, the element will be "
+        "p-refined."};
+  };
+
+  using options = tmpl::list<TargetTruncationError>;
+
+  static constexpr Options::String help = {
+      "Refine the grid towards the target truncation error"};
+
+  TruncationError() = default;
+
+  TruncationError(const double target_truncation_error)
+      : target_truncation_error_(target_truncation_error) {}
+
+  /// \cond
+  explicit TruncationError(CkMigrateMessage* msg) : Criterion(msg) {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(TruncationError);  // NOLINT
+  /// \endcond
+
+  using compute_tags_for_observation_box = tmpl::list<>;
+
+  using argument_tags = tmpl::list<::Tags::DataBox>;
+
+  template <typename DbTagsList, typename Metavariables>
+  auto operator()(const db::DataBox<DbTagsList>& box,
+                  Parallel::GlobalCache<Metavariables>& /*cache*/,
+                  const ElementId<Dim>& /*element_id*/) const {
+    auto result = make_array<Dim>(Flag::DoNothing);
```

Comment:
So the way this criterion is written, AMR will never be able to decrease the resolution as returning DoNothing will always take priority over any criteria that wants to DecreaseResolution.  So this criterion should handle when to request to decrease the resolution as well...

Reply:
Can I return `Undefined` to signal that this criterion makes no statement about what should happen in that dimension?

Reply:
How can this ever not have an opinion on what to do? The error is either above or below tolerances.

Reply:
Currently it doesn't try to coarsen the resolution, but I can add that. There should probably be a safety factor though or an acceptable absolute error range so it doesn't switch back and forth between resolutions at every AMR step?

Reply:
As a first pass I was going to ignore coarsening. We have to experiment with this anyway.

Reply:
Let's just have a meet about this. That's easier than going back and forth on gitHub comments :)

Reply:
As AMR is currently implemented, you should return `Join` (the lowest priority flag) if you do not want to override the decisions of other criteria.  But as this is a p-refinement criterion, it should return one of `{IncreaseResolution, DoNothing, DecreaseResolution}` I would have expected this criterion to compute a third number besides the error and target error, such as the estimated error dropping the top mode, in order to determine if the resolution can be decreased.